src/gwt/src/org/rstudio/studio/client/workbench/views/terminal/TerminalSession.java

```diff
@@ -0,0 +1,238 @@
+                                        TerminalDataInputEvent.Handler,
+                                        TerminalSessionStartedEvent.HasHandlers,
+                                        TerminalSessionStoppedEvent.HasHandlers
+{
+   public TerminalSession(final ShellSecureInput secureInput)
+   {
+      RStudioGinjector.INSTANCE.injectMembers(this);
+      secureInput_ = secureInput;
+      setHeight("100%");
+   }
+   
+   @Inject
+   private void initialize(WorkbenchServerOperations server)
+   {
+      server_ = server;
+   }
+   
+   /**
+    * Create a terminal process and connect to it.
+    */
+   public void connect()
+   {
+      server_.startShellDialog(ConsoleProcess.TerminalType.XTERM, 
+                               80, 25,
+                               false, /* not a modal dialog */
+                               new ServerRequestCallback<ConsoleProcess>()
+      {
+         @Override
+         public void onResponseReceived(ConsoleProcess consoleProcess)
+         {
+            consoleProcess_ = consoleProcess;
+            
+            if (getInteractionMode() != ConsoleProcessInfo.INTERACTION_ALWAYS)
+            {
+               writeError("Unsupported ConsoleProcess interaction mode");
+               return;
+            } 
+
+            if (consoleProcess_ != null)
+            {
+               addHandlerRegistration(consoleProcess_.addConsoleOutputHandler(TerminalSession.this));
+               addHandlerRegistration(consoleProcess_.addProcessExitHandler(TerminalSession.this));
+               addHandlerRegistration(addResizeTerminalHandler(TerminalSession.this));
+               addHandlerRegistration(addTerminalDataInputHandler(TerminalSession.this));
+
+               consoleProcess.start(new ServerRequestCallback<Void>()
+               {
+                  @Override
+                  public void onResponseReceived(Void response)
+                  {
+                     fireEvent(new TerminalSessionStartedEvent(terminalTitle_, TerminalSession.this));
+                  }
+                  
+                  @Override
+                  public void onError(ServerError error)
+                  {
+                     writeError(error.getUserMessage());
+                  }
+               });
+            }
+         }
+      
+         @Override
+         public void onError(ServerError error)
+         {
+            writeError(error.getUserMessage());
+         }
+         
+      });
+   }
+   
+   @Override
+   public void onConsoleOutput(ConsoleOutputEvent event)
+   {
+      write(event.getOutput());
+   }
+   
+   @Override
+   public void onProcessExit(ProcessExitEvent event)
+   {
+      unregisterHandlers();
+      if (consoleProcess_ != null)
+      {
+         consoleProcess_.reap(new VoidServerRequestCallback());
+      }
+     
+      consoleProcess_ = null;
+      fireEvent(new TerminalSessionStoppedEvent(terminalTitle_, this));
```

Comment:
Is there a reason that these events are not fired on the global event bus? We usually put events there by convention as it allows an easy extensibility point for other parts of the system to react. (It also makes registration/unregistration a little easier as you don't need `HasHandlers` or direct wiring.)